src/runtime/runtime_stm32f103xx.go

```diff
@@ -44,13 +44,145 @@ func initCLK() {
 	}
 }
 
+const tickMicros = 1000
+
+var (
+	timestamp        timeUnit // nanoseconds since boottime
+	timerLastCounter uint32   // 24 bits ticks
+)
+
+//go:volatile
+type isrFlag bool
+
+var timerWakeup isrFlag
+
+func initRTC() {
+	// Enable the PWR and BKP.
+	stm32.RCC.APB1ENR |= stm32.RCC_APB1ENR_PWREN | stm32.RCC_APB1ENR_BKPEN
+
+	// access to backup register
+	stm32.PWR.CR |= stm32.PWR_CR_DBP
+
+	// Enable LSE
+	stm32.RCC.BDCR |= stm32.RCC_BDCR_LSEON
+
+	// wait until LSE is ready
+	for stm32.RCC.BDCR&stm32.RCC_BDCR_LSERDY == 0 {
+	}
+
+	// Select LSE
+	stm32.RCC.BDCR |= stm32.RCC_RTCCLKSource_LSE
+
+	// set prescaler to "max" per datasheet
+	stm32.RTC.PRLH = stm32.RTC_PRLH_PRLH_Msk
+	stm32.RTC.PRLL = stm32.RTC_PRLL_PRLL_Msk
+
+	// set count to zero
+	stm32.RTC.CNTH = 0x0
+	stm32.RTC.CNTL = 0x0
+
+	// Enable RTC
+	stm32.RCC.BDCR |= stm32.RCC_BDCR_RTCEN
+
+	// Clear RSF
+	stm32.RTC.CRL &^= stm32.RTC_CRL_RSF
+
+	// Wait till flag is set
+	for stm32.RTC.CRL&stm32.RTC_CRL_RSF == 0 {
+	}
+}
+
+// Enable the TIM3 clock.
+func initTIM() {
+	stm32.RCC.APB1ENR |= stm32.RCC_APB1ENR_TIM3EN
+
+	arm.SetPriority(stm32.IRQ_TIM3, 0xc3)
+	arm.EnableIRQ(stm32.IRQ_TIM3)
+}
+
+// sleepTicks should sleep for specific number of microseconds.
 func sleepTicks(d timeUnit) {
-	// TODO: use a real timer here
-	for i := 0; i < int(d/535); i++ {
-		arm.Asm("")
+	for d != 0 {
+		ticks()            // update timestamp
+		ticks := uint32(d) // current scaling only supports 100 usec to 6553 msec
+		timerSleep(ticks)
+		d -= timeUnit(ticks)
 	}
 }
 
+// number of ticks (microseconds) since start.
 func ticks() timeUnit {
-	return 0 // TODO
+	// convert RTC counter from seconds to microseconds
+	timerCounter := uint32(stm32.RTC.CNTH<<16|stm32.RTC.CNTL) * 1000 * 1000
```

Comment:
Why not `uint64`? It would increase the range by a large amount. At the moment, you're left with only about 12 bits of range after the multiplication.
Note: 64-bit arithmetic is generally quite cheap on Cortex-M, only division is expensive (but division is always expensive unless it can be optimized to a bit shift).

Reply:
Done.